// strong-agent/lib/profilers/memory.js
function debug (format, args) {
  if (/memprof/.test(process.env.NODEFLY_DEBUG) ) {
    console.log.apply(console, ['MEMORY PROFILER: ' + format].concat(args || []));
  }
}

// Load dependencies
var _ = require('underscore');

function Instances () {
  this.nf = global.nodefly;
  this.enabled = false;
  this.instances = [];

  try {
    this.memwatch = require('memwatch');
  } catch (e) {
    console.log('memwatch must be installed to use the instances feature');
  }

  // NOTE: Can not be prototype function. Difficult to bind and use with off()
  var self = this;
  this._step = function () {
    debug('instance monitoring step');

    // Stop heap diff and get the change data
    self.diff.end().change.details.forEach(function (change) {
      // Attempt to find existing instance
      var inst = _(self.instances).find(function (inst) {
        return inst.type === change.what;
      });

      // If not present, create a new one from the current change item
      if (typeof inst === 'undefined') {
        inst = { type: change.what, total: 0, size: 0 };
        self.instances.push(inst);
      }

      // Adjust total
      inst.total += change['+'];
      inst.total -= change['-'];

      // Grab updated memory usage
      inst.size += change.size_bytes;
    });

    // Emit state
    self.nf.emit('instances', {
      type: 'Instances'
      , state: self.instances
    });
    
    // Reset heap diff    
    self.diff = new self.memwatch.HeapDiff();
  };
}
module.exports = new Instances;

Instances.prototype.init = function () {
  this.nf = global.nodefly;
};

Instances.prototype.toggle = function () {
  this.enabled ? this.stop() : this.start();
};

Instances.prototype.start = function () {
  if ( ! this.memwatch) return;
  debug('instance monitoring started');
  this.instances = [];
  this.diff = new this.memwatch.HeapDiff();
  this.timer = setInterval(this._step, 15 * 1000);
  // this.memwatch.on('stats', this._step);
  this._step();
  this.enabled = true;
};

Instances.prototype.stop = function () {
  if ( ! this.memwatch) return;
  debug('instance monitoring stopped');
  clearInterval(this.timer);
  // this.memwatch.removeListener('stats', this._step);
  this.enabled = false;
};
