// strong-agent/lib/loop.js
function debug (format, args) {
	if (/uvmon/.test(process.env.NODEFLY_DEBUG) ) {
		console.log.apply(console, ['UVMON: ' + format].concat(args || []));
	}
}

var nf;
var proxy = require('./proxy');
var util = require('util');
var uvmon;
try {
	uvmon = require('nodefly-uvmon');
} catch(e) {
	console.log('unable to load optional dependency nodefly-uvmon');
}

var config = global.nodeflyConfig;

exports.init = function() {
	nf = global.nodefly;
	if (!uvmon) return;
	start();
}

function nodeflyNoOp() {}

function checkNextTick(obj, args) {
	proxy.callback(args, -1, nodeflyNoOp);	
}

function checkTimers(obj, args){
	// callback for any setTimeout or setInterval
	proxy.callback(args, -1, nodeflyNoOp);
}


function start() {
	debug('starting uvmon');
	setInterval(function() {
		var data = {};
		data.loop = uvmon.getData();
		data._ts = nf.millis();
		nf.emit('loop', data);
		// data is automatically reset in getData()

		if (!data.loop) {
			data.loop = { slowest_ms: 0, sum_ms: 0, count: 1 };
			debug('uvmon did not report data, using dummy data');
		}
		else {
			debug('uvmon data: \n', [ util.inspect(data.loop) ]);
		}

		// we're also going to shoehorn it into the metric data to make our life easier
		nf.metric(null, 'queue', [data.loop.slowest_ms, (data.loop.sum_ms / data.loop.count)]);

		if (process.env.NODEFLY_DEBUG && /uvmon/.test(process.env.NODEFLY_DEBUG)) {
			console.error('UVMON: %s', JSON.stringify(data));
		}
	}, config.loopInterval);



	proxy.before(process, [ 'nextTick' ], checkNextTick);
	proxy.before(global, [ 'setTimeout', 'setInterval' ], checkTimers);

}
